File: day01/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	m := map[int]struct{}{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		num, err := strconv.Atoi(scanner.Text())
		if err != nil {
			log.Fatal(err)
		}
		k := 2020 - num
		if _, ok := m[k]; ok {
			fmt.Println("Part 1:", k*num)
		}
		m[num] = struct{}{}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
pt2loop:
	for i := range m {
		for j := range m {
			k := 2020 - i - j
			if _, ok := m[k]; ok {
				fmt.Println("Part 2:", i*j*k)
				break pt2loop
			}
		}
	}
}

File: day01/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let nums = {
        let mut nums = HashSet::new();
        for line in reader.lines() {
            nums.insert(line?.parse::<i32>()?);
        }
        nums
    };

    for i in &nums {
        let k = 2020 - i;
        if nums.contains(&k) {
            println!("Part 1: {}", i * k);
            break;
        }
    }

    'pt2loop: for i in &nums {
        for j in &nums {
            let k = 2020 - i - j;
            if nums.contains(&k) {
                println!("Part 2: {}", i * j * k);
                break 'pt2loop;
            }
        }
    }

    Ok(())
}
